Fixes for HTTP 1.0 not working when Content-Length not given

mcunha opened the pull request:
Any HTTP 1.0 requests where hanging when Content-Length (CL) was not provided due to ShouldKeepAlive being false and parser waiting for end of request body.

Fixes include:
- Parser should not wait for a request body [that may not arrive] when CL is missing and Transfer-Encoding (TE) are missing.
- Fix test to follow standard more closely, a POST request without CL or TE should not be accepted (HTTP 400 or 411), so Body should be null as it was ignored.

NOTE: Failure to provide CL is a violation of HTTP/1.0 and failure to provide CL _and_ TE is a violation of HTTP/1.1 and should result in HTTP 400 or 411.

mcunha committed:
Don't wait for request body unless really needed

Do not wait for a request body [that will never arrive] when
Content-Length (CL) is missing and Transfer-Encoding (TE) are missing.
Fix test to follow standard, a POST request without CL or TE should not
be accepted (HTTP 400 or 411).

bvanderveen commented:
Could you add a few test cases which would fail without the patch and pass with it?

On Mar 28, 2012, at 7:28, Marco Cunhareply@reply.github.com wrote:

> HTTP 1.0 requests where hanging when Content-Length was not provided due to ShouldKeepAlive being false and parser waiting for end of request body.
> 
> Fixes include:
> - Do not wait for a request body [that may not arrive] when Content-Length (CL) is missing and Transfer-Encoding (TE) are missing.
> - Fix test to follow standard more closely, a POST request without CL or TE should not be accepted (HTTP 400 or 411), so Body should be null as it was ignored.
> - Failure to provide CL is a violation of HTTP/1.0 and failure to provide CL AND TE is a violation of HTTP/1.1 and should result in HTTP 400 or 411.
> 
> You can merge this Pull Request by running:
> 
>  git pull https://github.com/mcunha/httpmachine master
> 
> Or you can view, comment on it, or merge it online at:
> 
>  https://github.com/bvanderveen/httpmachine/pull/6
> 
> -- Commit Summary --
> - Don't wait for request body unless really needed
> 
> -- File Changes --
> 
> M src/HttpMachine/HttpParser.cs (34)
> M src/HttpMachine/rl/HttpParser.cs.rl (14)
> M tests/HttpMachine.Tests/TestRequest.cs (2)
> 
> -- Patch Links --
> 
>  https://github.com/bvanderveen/httpmachine/pull/6.patch
>  https://github.com/bvanderveen/httpmachine/pull/6.diff
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/bvanderveen/httpmachine/pull/6

mcunha committed:
Add test for POST's without Content-Length

mcunha committed:
Forgot fixture for POST without Content-Length

mcunha committed:
Test harness should not provide EOF bump

Clients won't close connection until they receive one reply at least,
so test harness should not provide EOF marker to parser.

mcunha committed:
Ignore MonoDevelop tests results

mcunha commented:
Please ignore this pull request sending you another one in a sec.
